pkg/cmd/repo/clone/clone.go

```diff
@@ -82,44 +82,65 @@ func cloneRun(opts *CloneOptions) error {
 	}
 
 	apiClient := api.NewClientFromHTTP(httpClient)
-	cloneURL := opts.Repository
-	if !strings.Contains(cloneURL, ":") {
-		if !strings.Contains(cloneURL, "/") {
+
+	respositoryIsURL := strings.Contains(opts.Repository, ":")
+	repositoryIsFullName := !respositoryIsURL && strings.Contains(opts.Repository, "/")
+
+	var repo ghrepo.Interface
+	var protocol string
+	if respositoryIsURL {
+		repoURL, err := git.ParseURL(opts.Repository)
+		if err != nil {
+			return err
+		}
+		repo, err = ghrepo.FromURL(repoURL)
+		if err != nil {
+			return err
+		}
+		if repoURL.Scheme == "git+ssh" {
+			repoURL.Scheme = "ssh"
+		}
+		protocol = repoURL.Scheme
+	} else {
+		var fullName string
+		if repositoryIsFullName {
+			fullName = opts.Repository
+		} else {
 			currentUser, err := api.CurrentLoginName(apiClient, ghinstance.OverridableDefault())
 			if err != nil {
 				return err
 			}
-			cloneURL = currentUser + "/" + cloneURL
+			fullName = currentUser + "/" + opts.Repository
 		}
-		repo, err := ghrepo.FromFullName(cloneURL)
+
+		repo, err = ghrepo.FromFullName(fullName)
 		if err != nil {
 			return err
 		}
 
-		protocol, err := cfg.Get(repo.RepoHost(), "git_protocol")
+		protocol, err = cfg.Get(repo.RepoHost(), "git_protocol")
 		if err != nil {
 			return err
 		}
-		cloneURL = ghrepo.FormatRemoteURL(repo, protocol)
 	}
 
-	var repo ghrepo.Interface
-	var parentRepo ghrepo.Interface
-
-	// TODO: consider caching and reusing `git.ParseSSHConfig().Translator()`
-	// here to handle hostname aliases in SSH remotes
-	if u, err := git.ParseURL(cloneURL); err == nil {
-		repo, _ = ghrepo.FromURL(u)
+	// Load the repo from the API to get the username/repo name in its
+	// canonical capitalization
+	var canonicalRepo ghrepo.Interface
+	canonicalRepo, err = api.GitHubRepo(apiClient, repo)
+	if err != nil {
+		return err
 	}
+	canonicalCloneURL := ghrepo.FormatRemoteURL(canonicalRepo, protocol)
 
-	if repo != nil {
-		parentRepo, err = api.RepoParent(apiClient, repo)
-		if err != nil {
-			return err
-		}
+	cloneDir, err := git.RunClone(canonicalCloneURL, opts.GitArgs)
+	if err != nil {
+		return err
 	}
 
-	cloneDir, err := git.RunClone(cloneURL, opts.GitArgs)
+	// If the repo is a fork, add the parent as an upstream
+	var parentRepo ghrepo.Interface
+	parentRepo, err = api.RepoParent(apiClient, canonicalRepo)
```

Comment:
Since we're already making an API request in `GitHubRepo()` above, a potential improvement here could be to expand (or rather, duplicate) that query to also include its parent repository in the response. That way we could get the information about the repo + its parent in a single request.

You are welcome to amend this PR if you're willing to add this tiny optimization, but you can also leave it as-is and our team can get to this as well. 👍

Reply:
I've pushed a change that achieves this by loading the parent in the `GitHubRepo` query - it does mean grabbing a few more bytes than required if you don't care about the parent, but I thought it might be good practice because otherwise the parent (which was already in the model) would be nil - this could cause a bug by implying that the repo isn't a fork (when it is).